### PR TITLE
Runtime: Argo Bridge

### PR DESCRIPTION
**Pull request by freakstatic**
Still missing the benchmark tests

### CODE REVIEW

File: runtime-modules/argo-bridge/src/Untitled-2
Comment (kdembler) [inline runtime-modules/argo-bridge/src/Untitled-2]: This file shouldn't be committed

File: runtime-modules/argo-bridge/src/errors.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/errors.rs]: 1. `t` missing in `Insufficient`
2. `e` missing in `Bridge`

File: runtime-modules/argo-bridge/src/events.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/events.rs]: Let's include the thawn end block in the event

File: runtime-modules/argo-bridge/src/errors.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/errors.rs]: Let's update the comment after change to `thawn_ends_at`

File: runtime-modules/argo-bridge/src/lib.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/lib.rs:59]: The `production_config` in `chain_spec/argo_bridge_config.rs` contains a value for default bridging fee, are those 2 connected?

File: runtime-modules/argo-bridge/src/lib.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/lib.rs]: This should be `<=`

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/tests/mod.rs:56]: I'm not sure what kind of assertions we can do in our tests, but could this also assert:
1. Proper balance change in the caller account
2. Proper change of mint allowance?

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/tests/mod.rs]: Name of this test ends in `_success` but we assert error here. I think this should be the happy case? In that case, same questions as for the outbound transfer happy case - can we assert change in mint allowance and in token balances?

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (kdembler) [inline runtime-modules/argo-bridge/src/tests/mod.rs]: Can we assert proper updated status?

File: runtime/src/proposals_configuration/defaults.rs
Comment (kdembler) [inline runtime/src/proposals_configuration/defaults.rs]: Let's make this 3/3.

**Review by kdembler — CHANGES_REQUESTED** [review]

File: runtime-modules/argo-bridge/src/lib.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/lib.rs]: Here I would use as error message `OperatorAccountNotSet` (provide that we need at some point the operator account to be `None`)

File: runtime-modules/argo-bridge/src/lib.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/lib.rs:67]: Does it make sense to have a bridge without operator account? If not I would say to remove the optionality. So you can later avoid an extra ensure check

File: runtime-modules/argo-bridge/src/lib.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/lib.rs]: Same as above I will use as error variant name `OperatorAccountNotSet`

File: runtime-modules/argo-bridge/src/types.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/types.rs:36]:
```suggestion
    // EVM-based addresses only need 20 bytes but we use 32 bytes to enable compatibility with other chains
```

File: runtime-modules/argo-bridge/src/types.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/types.rs]:
```suggestion
    // When using EVM-based addresses, the last 12 bytes should be zero
```

File: runtime-modules/argo-bridge/src/types.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/types.rs:36]: Phantom, Polygon etc.. all have 20 bytes long addresses

File: runtime-modules/argo-bridge/src/errors.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/errors.rs:21]: As I said above I would add a `OperatorAccountNotSet` variant if really necessary here

File: runtime-modules/argo-bridge/src/types.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/types.rs:38]: I would use `address` here and not `account`

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/tests/mod.rs:85]: I will still add the line in order to update the sender balance. In case in a future refactoring we swap the line with the balance check and the bridge pausing check not to have this test failing

File: runtime-modules/argo-bridge/src/types.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/types.rs]: The most significant 12 bytes should be 0, this padding method is consistent with the way Solidity handles addresses and other data types, ensuring compatibility and consistency across the Ethereum ecosystem.
So the first 12 bytes should be 0 and not the last 12

Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/types.rs]: cc: @kdembler this decision might have consequence on how you parse events fields on the argo squid

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/tests/mod.rs:461]: There are the following error-path case missing:
`reqquest_outbound_transfer`:
  - remote chain not supported
`finish_pause_bridge`:
  - operator account is `None` case

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/tests/mod.rs:56]: Also proper event deposit (very important for the argo-squid)

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/tests/mod.rs]: also check proper event deposit

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/tests/mod.rs]: and event deposit

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/tests/mod.rs:388]: also check for event deposit

**Review by dobertRowneySr — COMMENTED** [review]
The pallet looks solid, the only issues that I have found are:
- remote chain address formatting : 12 most significant bytes should be zero IMHO
- If you can provide a case where the operator account is `None`
The tests needs some touching

**Review by dobertRowneySr — CHANGES_REQUESTED** [review]
The pallet looks solid, the only issues that I have found are:
- remote chain address formatting : 12 most significant bytes should be zero IMHO
- If you can provide a case where the operator account is `None`
The tests needs some touching, especially in the happy path assertions (especially on event production)
Also: wouldn't it be easier for you to have multiple commits 😅 , instead of a single huge one ?

**Review by dobertRowneySr — COMMENTED** [review]
Ignore this, I accidentally re requested a review

File: bin/node/src/chain_spec/argo_bridge_config.rs
Comment (dobertRowneySr) [inline bin/node/src/chain_spec/argo_bridge_config.rs:3]: I also think that you can set the mint allowance storage variable to be `0` by default since I don't think any other initial value is plausible

File: runtime/src/lib.rs
Comment (dobertRowneySr) [inline runtime/src/lib.rs:146]: petra should be spec version `2004`, correctly if I am wrong @kdembler

File: runtime-modules/argo-bridge/src/lib.rs
Comment (dobertRowneySr) [inline runtime-modules/argo-bridge/src/lib.rs]: I believe you can set default value of this to `0` if I correctly understood the bridging accounting logic, as there shouldn't be any other possible initial value

File: runtime/src/lib.rs
Comment (dobertRowneySr) [inline runtime/src/lib.rs]: 😅 1k USD for bridging, shouldn't this be lower?

**Review by dobertRowneySr — CHANGES_REQUESTED** [review]
Did another quick review round

**Comment by freakstatic** [comment]
> The pallet looks solid, the only issues that I have found are:
> 
> * remote chain address formatting : 12 most significant bytes should be zero IMHO
> * If you can provide a case where the operator account is `None`
>   The tests needs some touching

Thanks for the review :+1: 
The operator account being `None` is because I wasn't sure which account should be the default one :thinking:

File: runtime-modules/argo-bridge/src/lib.rs
Comment (kdembler) [reply runtime-modules/argo-bridge/src/lib.rs:67]: Not sure what we would set in the initial pallet config in that case?

File: runtime-modules/argo-bridge/src/types.rs
Comment (kdembler) [reply runtime-modules/argo-bridge/src/types.rs:36]: well they're still Ethereum addresses 😅

File: runtime-modules/argo-bridge/src/lib.rs
Comment (freakstatic) [reply runtime-modules/argo-bridge/src/lib.rs]: in the tests is handy to be able to set a value for the mint allowance in the config

Comment (freakstatic) [reply runtime-modules/argo-bridge/src/lib.rs]: or were you really just suggesting a default value and not removing it from the genesis config?

File: runtime-modules/argo-bridge/src/lib.rs
Comment (freakstatic) [reply runtime-modules/argo-bridge/src/lib.rs:59]: yes but I was missing the call to `DefaultBridgingFee::get()`

File: runtime-modules/argo-bridge/src/tests/mod.rs
Comment (freakstatic) [reply runtime-modules/argo-bridge/src/tests/mod.rs]: > Name of this test ends in `_success` but we assert error here. I think this should be the happy case? In that case, same questions as for the outbound transfer happy case - can we assert change in mint allowance and in token balances?

the title was good but the test was wrong...
added checks for the balance and the event deposit

File: runtime-modules/argo-bridge/src/types.rs
Comment (kdembler) [reply runtime-modules/argo-bridge/src/types.rs]: Yeah that's what I meant actually, my mistake writing spec. I suggest
```suggestion
    // When using EVM addresses, the 12 most-significant bytes should be zero
```
Squid already uses this approach

File: runtime-modules/argo-bridge/src/types.rs
Comment (kdembler) [reply runtime-modules/argo-bridge/src/types.rs:38]: Yeah I spent a minute on this. My thinking was that this could also be a Substrate chain in the future and in that case it would actually be 32-byte AccountId, so account looked more universal, I don't mind either way

File: runtime-modules/argo-bridge/src/lib.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/lib.rs:67]: Ok so the actual procedure for deployment will be:
1. Deploy petra with argo pallet
2. Setup JoyOp account signatories
3. Activate the bridge

If so Then that's ok and the operator account can be none

**Comment by dobertRowneySr** [comment]
All my previous comments have been addressed. So for now LGTM
Should I wait for the benchmarks?

**Comment by freakstatic** [comment]
@ignazio-bovo and @kdembler I have added the benchmark tests please take a look.
the weight calculation is not finding the benchmarks so I'm most likely missing something...

`./scripts/../target/debug/joystream-node benchmark pallet --pallet=argo_bridge --extrinsic="*" --chain=prod-test --steps=50 --repeat=20 --execution=wasm --template=./scripts/../devops/joystream-pallet-weight-template.hbs --output=./scripts/../runtime-modules/argo-bridge/src/weights.rs`

```
2024-05-30 05:28:07 [0] 💸 generated 1 npos voters, 1 from validators and 0 nominators    
2024-05-30 05:28:07 [0] 💸 generated 1 npos targets    
Error: Input("No benchmarks found which match your input.")
```

File: runtime/src/lib.rs
Comment (kdembler) [reply runtime/src/lib.rs:146]: @freakstatic please bump

File: runtime/src/lib.rs
Comment (kdembler) [inline runtime/src/lib.rs]: Let's try placing the ArgoBridge pallet last, after Proxy. This way we should not change existing modules indexes and won't need to bump `transaction_version` (only `spec_version`)

**Comment by kdembler** [comment]
@ignazio-bovo @freakstatic I think the discussion on storage initialization is still not resolved. @ignazio-bovo could you expand on what you had in mind?

**Comment by mnaamani** [comment]
> @ignazio-bovo and @kdembler I have added the benchmark tests please take a look. the weight calculation is not finding the benchmarks so I'm most likely missing something...
> 
> `./scripts/../target/debug/joystream-node benchmark pallet --pallet=argo_bridge --extrinsic="*" --chain=prod-test --steps=50 --repeat=20 --execution=wasm --template=./scripts/../devops/joystream-pallet-weight-template.hbs --output=./scripts/../runtime-modules/argo-bridge/src/weights.rs`
> 
> ```
> 2024-05-30 05:28:07 [0] 💸 generated 1 npos voters, 1 from validators and 0 nominators    
> 2024-05-30 05:28:07 [0] 💸 generated 1 npos targets    
> Error: Input("No benchmarks found which match your input.")
> ```

I think you need to add the argo bridge pallet to the `define_benchmarks!` macro in `runtime/src/runtime_api.rs`

```diff
diff --git a/runtime/src/runtime_api.rs b/runtime/src/runtime_api.rs
index 8d3a06316a..c57e968e51 100644
--- a/runtime/src/runtime_api.rs
+++ b/runtime/src/runtime_api.rs
@@ -127,6 +127,7 @@ mod benches {
         [content, Content]
         [project_token, ProjectToken]
         [pallet_proxy, Proxy]
+        [argo_bridge, ArgoBridge]
     );
 }
```

**Comment by ignazio-bovo** [comment]
> @ignazio-bovo @freakstatic I think the discussion on storage initialization is still not resolved. @ignazio-bovo could you expand on what you had in mind?

Sure: 
Originally I though that when the argo pallet will be release in the next runtime upgrade, it would be activated immediately -> hence at upgrade time we will already have JoyOp account setup with proper multisig scheme and signatories. If you are thinking about this then the JoyOp account is established before the argo pallet is released and it can be included in the initial chainspec.

If our #1 goal is to release the runtime upgrade first in order to have the argo on runtime and activate the bridge any time later then and keep it paused in the meantime then it makes sense to allow the JoyOp account to be `None` and set it later on before setting the bridge status to active

File: runtime-modules/argo-bridge/src/lib.rs
Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/lib.rs]: the storage value mint_allowance is increased during Outbound transfer and decreased during Inbound transfer and it represent the total amount of JOY we have outside of the Joystream mainnet network.
If so my observation was: since the only plausible initial value is zero, we can set the default value when declearing the storage variable inside the decl_storage macro instead of manually setting the value in the chainspec config. (Even though I remember with confidence that in case a storage variable is not initialised it will be set to the fault value when accessed for the first time, so 0 for mint allowance which in production will be a u128)

Comment (dobertRowneySr) [reply runtime-modules/argo-bridge/src/lib.rs]: this is a minor comment however, I am fine even with the existing solution

**Comment by freakstatic** [comment]
> > @ignazio-bovo and @kdembler I have added the benchmark tests please take a look. the weight calculation is not finding the benchmarks so I'm most likely missing something...
> > `./scripts/../target/debug/joystream-node benchmark pallet --pallet=argo_bridge --extrinsic="*" --chain=prod-test --steps=50 --repeat=20 --execution=wasm --template=./scripts/../devops/joystream-pallet-weight-template.hbs --output=./scripts/../runtime-modules/argo-bridge/src/weights.rs`
> > ```
> > 2024-05-30 05:28:07 [0] 💸 generated 1 npos voters, 1 from validators and 0 nominators    
> > 2024-05-30 05:28:07 [0] 💸 generated 1 npos targets    
> > Error: Input("No benchmarks found which match your input.")
> > ```
> 
> I think you need to add the argo bridge pallet to the `define_benchmarks!` macro in `runtime/src/runtime_api.rs`
> 
> ```diff
> diff --git a/runtime/src/runtime_api.rs b/runtime/src/runtime_api.rs
> index 8d3a06316a..c57e968e51 100644
> --- a/runtime/src/runtime_api.rs
> +++ b/runtime/src/runtime_api.rs
> @@ -127,6 +127,7 @@ mod benches {
>          [content, Content]
>          [project_token, ProjectToken]
>          [pallet_proxy, Proxy]
> +        [argo_bridge, ArgoBridge]
>      );
>  }
> ```

thanks very much 🙏 
yeah that was exactly what I was missing

**Comment by freakstatic** [comment]
@ignazio-bovo and @kdembler  can you please review the benchmarking tests?

File: runtime-modules/proposals/codex/src/benchmarking.rs
Comment (freakstatic) [inline runtime-modules/proposals/codex/src/benchmarking.rs]: this is still WIP

**Review by dobertRowneySr — APPROVED** [review]
Benchmarks looks ok in terms of worst case scenario accounting

**Review by dobertRowneySr — CHANGES_REQUESTED** [review]
There are some errors that needs to be fixed (see CI).
You can test the benchmarking code with cargo test -p "pallet-argo-bridge" --features "runtime-benchmarks" (you can also use clippy with the same feature enabled)

**Comment by kdembler** [comment]
@freakstatic @ignazio-bovo

I opened a PR with fixes for clippy issues that we should merge into this PR: https://github.com/freakstatic/joystream/pull/2

I have also opened a new PR based on this to add a revert extrinsic: https://github.com/Joystream/joystream/pull/5158

Please check

**Review by kdembler — APPROVED** [review]
LGTM pending benchmark&weights approval from @ignazio-bovo

We also need to add overflow protection in some extrinsics but will do in a subsequent PR

**Review by dobertRowneySr — APPROVED** [review]
LGTM, the mac osx checks will be fixed in my PR here: #5157

**Pull request merged**
